File: good.js
const fs = require('fs');
const pptxgen = require('pptxgenjs');
const { parse } = require('svgson');
const colornames = require('colornames'); // カラーネーム変換

// SVGファイルのパス
const svgFilePath = "1.svg";

// PowerPointのスライドサイズ（デフォルト: 10 × 5.625 インチ）
const pptWidth = 10;
const pptHeight = 5.625;

// デバッグ用ログ
const DEBUG = false;

// SVGファイルの読み込み
fs.readFile(svgFilePath, 'utf8', (err, data) => {
    if (err) {
        console.error("SVGファイルの読み込みエラー:", err);
        return;
    }

    // SVGの解析
    parse(data).then(svgJson => {
        if (DEBUG) console.log("SVG要素:", JSON.stringify(svgJson, null, 2));

        const pptx = new pptxgen();
        let slide = pptx.addSlide();
        let rects = [];

        const viewBox = svgJson.attributes.viewBox ? svgJson.attributes.viewBox.split(" ") : [0, 0, 1600, 900];
        const [vbX, vbY, vbWidth, vbHeight] = viewBox.map(Number);

        // スケール変換係数 (SVG → PPT)
        const scaleX = pptWidth / vbWidth;
        const scaleY = pptHeight / vbHeight;
        const scale = Math.min(scaleX, scaleY); // 等比スケール

        // 色変換関数（RGBA & colornames 対応）
        function convertColor(color) {
            if (!color) return "000000"; // デフォルトの黒
            if (colornames(color)) return colornames(color).replace("#", "").toUpperCase();
            if (color.startsWith("rgba")) {
                let rgba = color.match(/([\d.]+)/g);
                if (rgba && rgba.length === 4) {
                    let hex = (
                        ((1 << 24) + (parseInt(rgba[0]) << 16) + (parseInt(rgba[1]) << 8) + parseInt(rgba[2]))
                        .toString(16)
                        .slice(1)
                    );
                    return hex.toUpperCase();
                }
            }

            if (color.startsWith("url(")) return "2563EB"; // グラデーションは青 (#2563EB) に統一
            return color.replace("#", "").toUpperCase(); // PPTX用のRGBフォーマット
        }

        // SVG要素をPPTXに変換する関数
        function processElement(element, parentTransform = { x: 0, y: 0 }) {
            const { name, attributes, children } = element;
            let x = (parseFloat(attributes.x || 0) - vbX + parentTransform.x) * scale;
            let y = (parseFloat(attributes.y || 0) - vbY + parentTransform.y) * scale;
            let width = parseFloat(attributes.width || 100) * scale;
            let height = parseFloat(attributes.height || 30) * scale;
            let rx = parseFloat(attributes.rx || 0) * scale;
            let ry = parseFloat(attributes.ry || 0) * scale;
            let fill = convertColor(attributes.fill || "#FFFFFF");
            let opacity = parseFloat(attributes.opacity || attributes["fill-opacity"] || 1);
        
            let fontSize = attributes["font-size"] ? parseFloat(attributes["font-size"].replace("px", "")) : 14;
            let textColor = convertColor(attributes.fill || "#000000");
            
            
            // **🔹 rect の処理**
            if (name === "rect") {
                let rectOptions = {
                    x: x,
                    y: y,
                    w: width,
                    h: height,
                    fill: { color: fill, transparency: 100 - (opacity * 100) }
                };

                // **スケール変換を適用**
                let rxScaled = rx * scale;
                let ryScaled = ry * scale;

                // **過度な丸みを抑制**
                let maxRadius = Math.min(width * 0.2, height * 0.2); // 最大でも幅・高さの20%まで
                let adjustedRx = Math.min(rxScaled, maxRadius);
                let adjustedRy = Math.min(ryScaled, maxRadius);
                let finalRadius = Math.max(adjustedRx, adjustedRy); // より大きい方を採用

                // **角丸の処理**
                if (finalRadius > 0) {
                    rectOptions.radius = finalRadius; // 適切な丸みを適用
                    slide.addShape(pptx.ShapeType.roundRect, rectOptions);
                } else {
                    slide.addShape(pptx.ShapeType.rect, rectOptions);
                }
            }

            // テキストの処理
            if (name === "text") {
                let textRuns = [];
                let totalText = ""; // すべてのテキストを統合


                children.forEach(child => {
                    if (child.name === "tspan") {
                        let tspanColor = convertColor(child.attributes.fill || textColor);
                        let tspanText = child.children.map(c => c.value || "").join("").trim();
                        let isBold = child.attributes["font-weight"] === "bold"; // 太字判定

                        totalText += tspanText;
                        if (tspanText) {
                            textRuns.push({
                                text: tspanText,
                                options: { color: tspanColor, bold: isBold }
                            });
                        }
                    } else {
                        let normalText = child.value || "";
                        let isBold = attributes["font-weight"] === "bold";

                        totalText += normalText;
                        if (normalText.trim()) {
                            textRuns.push({
                                text: normalText,
                                options: { color: textColor, bold: isBold }
                            });
                        }
                    }
                });

                if (textRuns.length > 0) {
                    let fontSizePx = fontSize;
                    let fontSizePt = fontSizePx * 0.75; // px → pt変換

                    let textBoxHeight = fontSizePt * 0.0352778; // pt → cm
                    let textBoxWidth = fontSizePt * totalText.length * 0.7 * 0.0352778; // 文字数に基づく幅

                    // **修正: `text` の `x, y` のスケール調整**
                    let textX = (parseFloat(attributes.x || 0) - vbX + parentTransform.x) * scaleX;
                    let textY = (parseFloat(attributes.y || 0) - vbY + parentTransform.y) * scaleY;

                    // **修正: ベースライン補正**
                    textY -= textBoxHeight * 0.35; // ベースライン基準を PowerPoint 仕様に補正

                    // **修正: text-anchor の影響を考慮したX座標調整**
                    let textAlign = "left";
                    let xOffset = 0;

                    if (attributes["text-anchor"] === "middle") {
                        textAlign = "center";
                        xOffset = -textBoxWidth / 2;
                    } else if (attributes["text-anchor"] === "end") {
                        textAlign = "right";
                        xOffset = -textBoxWidth;
                    }

                    let correctedX = textX + xOffset;
                    let correctedY = textY;

                    // **修正: PowerPointの `align` を適用**
                    slide.addText(textRuns, {
                        x: correctedX,
                        y: correctedY,
                        fontSize: fontSizePt,
                        w: textBoxWidth,
                        h: textBoxHeight,
                        autoFit: true,
                        align: textAlign
                    });
                }
            }

            
            // **🔹 circle の処理**
            else if (name === "circle") {
                let cx = parseFloat(attributes.cx || 0) - vbX;
                let cy = parseFloat(attributes.cy || 0) - vbY;
                let r = parseFloat(attributes.r || 10);

                slide.addShape(pptx.ShapeType.ellipse, {
                    x: (cx + parentTransform.x) * scale, 
                    y: (cy + parentTransform.y) * scale, 
                    w: r * 2 * scale, 
                    h: r * 2 * scale,
                    fill: { color: fill, transparency: 100 - opacity * 100 }
                });
            }

            // **🔹 polygon の処理**
            else if (name === "polygon") {
                let points = attributes.points.split(" ").map(p => {
                    let [px, py] = p.split(",").map(Number);
                    return { x: px * scale, y: py * scale };
                });

                slide.addShape(pptx.ShapeType.triangle, {
                    x, y, w: width, h: height, fill: { color: fill }
                });
            }

            // **🔹 グループ <g> の処理**
            else if (name === "g") {
                let groupTransform = { x, y };
                children.forEach(child => processElement(child, groupTransform));
                return;
            }

            if (children && children.length > 0) {
                children.forEach(child => processElement(child, parentTransform));
            }
        }

        svgJson.children.forEach(element => processElement(element));

        
        pptx.writeFile({ fileName: "presentation.pptx" }).then(() => {
            console.log("PPTXファイルが作成されました");
        });
    });
});

File: good.py
import xmltodict
from pptx import Presentation
from pptx.util import Inches, Pt
from pptx.dml.color import RGBColor
import re
import webcolors

# ✅ SVGファイルのパス
svg_file_path = "1.svg"

# ✅ PowerPointのスライドサイズ (16:9)
ppt_width = Inches(10)
ppt_height = Inches(5.625)

# ✅ デバッグ用フラグ
DEBUG = True

# ✅ SVGファイルの読み込み
with open(svg_file_path, "r", encoding="utf-8") as f:
    svg_data = f.read()
print("✅ SVG ファイルを読み込みました")

# ✅ SVGの解析
svg_json = xmltodict.parse(svg_data)

# ✅ viewBox の取得
viewBox = svg_json["svg"].get("@viewBox", "0 0 1600 900").split()
vbX, vbY, vbWidth, vbHeight = map(float, viewBox)
print(f"✅ viewBox: {vbX} {vbY} {vbWidth} {vbHeight}")

# ✅ スケール変換係数 (SVG → PPT)
scale_x = ppt_width / vbWidth
scale_y = ppt_height / vbHeight
scale = min(scale_x, scale_y)  # **等比スケールを適用**

# ✅ 色変換関数
def convert_color(color):
    if not color:
        return RGBColor(0, 0, 0)  # デフォルト黒
    if color.startswith("#"):
        return RGBColor(int(color[1:3], 16), int(color[3:5], 16), int(color[5:7], 16))
    try:
        rgb = webcolors.name_to_rgb(color)
        return RGBColor(rgb.red, rgb.green, rgb.blue)
    except ValueError:
        return RGBColor(37, 99, 235)  # **デフォルトは青 (#2563EB)**

# ✅ PowerPointの作成
prs = Presentation()
prs.slide_width = ppt_width
prs.slide_height = ppt_height
slide = prs.slides.add_slide(prs.slide_layouts[6])  # **白紙スライド**

# ✅ SVG要素をPPTXに変換する関数
def process_element(element, parent_transform=(0, 0)):
    if not isinstance(element, dict):
        return

    for tag, content in element.items():
        if not isinstance(content, dict):
            continue  # **無効な要素をスキップ**

        attrib = content.get("@", {})

        if DEBUG:
            print(f"🔹 検出: {tag} (属性: {attrib})")

        # **TEXT要素の処理**
        if tag == "text":
            text = content.get("#text", "").strip()
            if text:
                x = (float(attrib.get("@x", 0)) - vbX + parent_transform[0]) * scale
                y = (float(attrib.get("@y", 0)) - vbY + parent_transform[1]) * scale
                font_size = float(attrib.get("@font-size", "14").replace("px", "")) * 0.75
                text_color = convert_color(attrib.get("@fill", "#000000"))
                text_anchor = attrib.get("@text-anchor", "start")

                print(f"📌 TEXT 要素: \"{text}\"")
                print(f"➡ 座標: ({x:.2f}, {y:.2f}), サイズ: {font_size:.2f}pt, 色: {attrib.get('@fill', '#000000')}")

                textbox = slide.shapes.add_textbox(Inches(x), Inches(y), Inches(4), Inches(0.5))
                tf = textbox.text_frame
                tf.word_wrap = False
                p = tf.add_paragraph()
                p.text = text
                p.font.size = Pt(font_size)
                p.font.color.rgb = text_color

                if text_anchor == "middle":
                    textbox.left = Inches(x - 2)
                elif text_anchor == "end":
                    textbox.left = Inches(x - 4)

                print(f"✅ PowerPoint に追加: \"{text}\"")

        # **RECT要素の処理**
        elif tag == "rect":
            x = (float(attrib.get("@x", 0)) - vbX + parent_transform[0]) * scale
            y = (float(attrib.get("@y", 0)) - vbY + parent_transform[1]) * scale
            width = float(attrib.get("@width", "100")) * scale
            height = float(attrib.get("@height", "30")) * scale
            fill = convert_color(attrib.get("@fill", "#FFFFFF"))

            slide.shapes.add_shape(1, Inches(x), Inches(y), Inches(width), Inches(height)).fill.fore_color.rgb = fill
            print(f"✅ RECT追加: x={x:.2f}, y={y:.2f}, w={width:.2f}, h={height:.2f}, 色={attrib.get('@fill', '#FFFFFF')}")

        # **CIRCLE要素の処理**
        elif tag == "circle":
            cx = (float(attrib.get("@cx", 0)) - vbX + parent_transform[0]) * scale
            cy = (float(attrib.get("@cy", 0)) - vbY + parent_transform[1]) * scale
            r = float(attrib.get("@r", "10")) * scale
            fill = convert_color(attrib.get("@fill", "#000000"))

            slide.shapes.add_shape(1, Inches(cx - r), Inches(cy - r), Inches(r * 2), Inches(r * 2)).fill.fore_color.rgb = fill
            print(f"✅ CIRCLE追加: cx={cx:.2f}, cy={cy:.2f}, r={r:.2f}, 色={attrib.get('@fill', '#000000')}")

        # **G要素の処理（グループ要素）**
        elif tag == "g":
            new_transform = (float(attrib.get("@x", 0)) + parent_transform[0], float(attrib.get("@y", 0)) + parent_transform[1])
            process_element(content, new_transform)

        # **子要素の処理**
        process_element(content, parent_transform)

# ✅ SVG要素の変換
svg_elements = svg_json["svg"]
process_element(svg_elements)

# ✅ PPTXの保存
prs.save("presentation_fixed.pptx")
print("✅ PPTXファイルが作成されました: presentation_fixed.pptx")
